lib/ex_machina/ecto_strategy.ex

```diff
@@ -70,40 +59,76 @@ defmodule ExMachina.EctoStrategy do
     end
   end
 
-  defp cast_all_assocs(%{__struct__: schema} = struct) do
-    assocs = get_schema_assocs(schema)
+  defp cast_all_embeds(%{__struct__: schema} = struct) do
+    embed_keys = schema_embeds(schema)
 
-    Enum.reduce(assocs, struct, fn(assoc, struct) ->
-      casted_value = struct |> Map.get(assoc) |> cast_assoc(assoc, struct)
+    Enum.reduce(embed_keys, struct, fn(embed_key, struct) ->
+      casted_value = Map.get(struct, embed_key) |> cast_embed(embed_key, struct)
 
-      Map.put(struct, assoc, casted_value)
+      Map.put(struct, embed_key, casted_value)
     end)
   end
 
-  defp cast_assoc(original_assoc, assoc_key, %{__struct__: schema} = struct) do
-    case original_assoc do
-      has_or_embeds_many when is_list(has_or_embeds_many) ->
-        Enum.map(has_or_embeds_many, &(cast_assoc(&1, assoc_key, struct)))
+  defp cast_embed(embeds_many, embed_key, struct) when is_list(embeds_many) do
+    Enum.map(embeds_many, &(cast_embed(&1, embed_key, struct)))
+  end
+  defp cast_embed(embed, embed_key, %{__struct__: schema}) do
+    case embed do
+      %{} ->
```

Comment:
Maybe this could be simplified using an `if`

```elixir
if embed
  # do stuff
end
```

Reply:
👍 updated in b2ec4ca002be6d5a91609b45a4582e4652573e8d